src/BalanceTx.purs

```diff
@@ -270,72 +301,70 @@ data BuildTxRawFailureReason = CannotBuildTxRaw
--- TxBody. I've ignored validity interval too.
-newtype UnbalancedTransaction = UnbalancedTransaction
-  { unbalancedTx :: Transaction
-  , utxoIndex :: UtxoM
-  -- utxoIndex will include utxos for scripts as well presumably.
-  }
-
--- There is an estimate function here:
--- https://input-output-hk.github.io/cardano-node/cardano-api/src/Cardano.Api.Fees.html#estimateTransactionFee
--- Otherwise, we'd need a Haskell server to run the "exact" version, since we
--- can't run the Plutus interpreter via Purescript.
-calculateMinFee :: Transaction -> Either CalculateMinFeeFailure BigInt
-calculateMinFee = undefined
-
-signTx :: Transaction -> Aff Transaction
-signTx = undefined
+calculateMinFee'
+  :: Transaction
+  -> QueryM (Either CalculateMinFeeFailure BigInt)
+calculateMinFee' = calculateMinFee >>> map (bimap wrap unwrap)
 
 --------------------------------------------------------------------------------
 -- Balancing functions and helpers
 --------------------------------------------------------------------------------
 -- https://github.com/mlabs-haskell/bot-plutus-interface/blob/master/src/BotPlutusInterface/PreBalance.hs#L54
+-- FIX ME: UnbalancedTx contains requiredSignatories which would be a part of
+-- multisig but we don't have such functionality ATM.
 -- | Balances an unbalanced transaction
-balanceTxM
-  :: Address
-  -> UnbalancedTransaction
-  -> QueryM (Either BalanceTxFailure Transaction)
-balanceTxM ownAddr (UnbalancedTransaction { unbalancedTx, utxoIndex }) = do
-  utxos' :: Either BalanceTxFailure Utxo <-
-    utxosAt ownAddr <#>
-      note (UtxosAtFailure' $ wrap CouldNotGetUtxos) >>> map unwrap
-  case utxos' of
-    Left err -> pure $ Left err
-    Right utxos -> do
-      let -- Combines utxos at the user address and those from any scripts involved
-        -- with the contract.
-        utxoIndex' :: Utxo
-        utxoIndex' = utxos `Map.union` unwrap utxoIndex
-      -- Sign before instead of recursively signing in loop.
-      signedUnbalancedTx :: Transaction <- lift $ signTx unbalancedTx
-      prebalancedTx' :: Either BalanceTxFailure Transaction <-
-        loop utxoIndex' ownAddr [] signedUnbalancedTx
-      pure do
-        prebalancedTx :: Transaction <- prebalancedTx'
-        lmap ReturnAdaChangeFailure' $
-          returnAdaChange ownAddr utxoIndex' prebalancedTx
+balanceTxM :: UnbalancedTx -> QueryM (Either BalanceTxFailure Transaction)
+balanceTxM (UnbalancedTx { transaction: unbalancedTx, utxoIndex }) = do
+  ownAddr' <- getWalletAddress
+  collateral' <- getWalletCollateral
+  -- Saves on an extra nested case:
+  case ownAddr', collateral' of
+    Nothing, _ ->
+      pure $ Left $ GetWalletAddressFailure' $ wrap CouldNotGetNamiWalletAddress
+    _, Nothing ->
+      pure $ Left $ GetWalletCollateralFailure' $ wrap CouldNotGetNamiCollateral
+    Just ownAddr, Just collateral -> do
+      utxos' <- map unwrap <$> utxosAt ownAddr
+      utxoIndex''' <- liftEffect $ utxoIndexToUtxo utxoIndex
```

Comment:
If you're going to use several primes in a row, I might like to suggest using them in ascending order. Here we have `utxoIndex` to `utxoIndex'''`, to `utxoIndex''`